File: openvox/modules/roles/files/services/healthreport/app/collectors/ntfy_updates.py
"""Pending container image updates.

Diun already watches every image and publishes to ntfy on a 6h schedule
(services/diun/diun.yml). Reading its topic back costs nothing and avoids a
second registry-polling implementation.
"""

import json

from ..model import CollectorResult, Observation
from .base import collector, http_get


@collector("updates")
def collect(config, rules):
    result = CollectorResult(name="updates")
    obs = result.observations

    response = http_get(
        "%s/docker-updates/json" % config.ntfy_url.rstrip("/"),
        params={"poll": "1", "since": "%dh" % config.lookback_hours},
        timeout=30,
    )

    # ntfy streams newline-delimited JSON, one message per line.
    messages = []
    for line in response.text.splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            message = json.loads(line)
        except ValueError:
            continue
        if message.get("event") != "message":
            continue
        messages.append({
            "title": message.get("title", ""),
            "message": (message.get("message") or "")[:300],
            "time": message.get("time"),
        })

    result.data = {"count": len(messages), "messages": messages[:50]}
    obs.append(Observation(
        id="image_updates..",
        collector="updates",
        subject="homelab",
        kind="image_updates",
        value=len(messages),
        unit="images",
        message="%d container image update notifications in the last %dh"
                % (len(messages), config.lookback_hours),
        evidence={"topic": "docker-updates"},
    ))
    return result

File: tools/internal/healthreport/collectors/ntfy_updates.go
// Pending container image updates. Diun already watches every image and
// publishes to ntfy on a 6h schedule (services/diun/diun.yml). Reading its
// topic back costs nothing and avoids a second registry-polling
// implementation.
package collectors

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net/url"
	"strings"
	"time"

	hr "github.com/MrCodeEU/homelab-automation/tools/internal/healthreport"
)

func init() {
	hr.RegisterCollector("updates", collectUpdates)
}

func collectUpdates(cfg hr.Config, rules hr.RulesFile) *hr.CollectorResult {
	result := hr.NewCollectorResult("updates")

	target := fmt.Sprintf("%s/docker-updates/json?%s", strings.TrimRight(cfg.NtfyURL, "/"), url.Values{
		"poll": {"1"}, "since": {fmt.Sprintf("%dh", cfg.LookbackHours)},
	}.Encode())
	resp, err := httpGet(target, 30*time.Second, nil)
	if err != nil {
		panic(err)
	}
	defer resp.Body.Close()

	// ntfy streams newline-delimited JSON, one message per line.
	var messages []hr.UpdateMessage
	scanner := bufio.NewScanner(resp.Body)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var msg map[string]any
		if err := json.Unmarshal([]byte(line), &msg); err != nil {
			continue
		}
		if event, _ := msg["event"].(string); event != "message" {
			continue
		}
		title, _ := msg["title"].(string)
		body, _ := msg["message"].(string)
		if len(body) > 300 {
			body = body[:300]
		}
		messages = append(messages, hr.UpdateMessage{Title: title, Message: body, Time: msg["time"]})
	}

	sample := messages
	if len(sample) > 50 {
		sample = sample[:50]
	}
	result.Data = &hr.UpdatesData{Count: len(messages), Messages: sample}
	result.Observations = append(result.Observations, &hr.Observation{
		ID: "image_updates..", Collector: "updates", Subject: "homelab", Kind: "image_updates",
		Value: len(messages), Unit: "images",
		Message:  fmt.Sprintf("%d container image update notifications in the last %dh", len(messages), cfg.LookbackHours),
		Evidence: map[string]any{"topic": "docker-updates"}, Severity: "info",
	})
	return result
}
